process.go

```
@@ -0,0 +1,142 @@
+	}
+
+	// read directory
+	entries, err := ioutil.ReadDir(input)
+	if err != nil {
+		return err
+	}
+
+	// process or dive in again
+	for _, entry := range entries {
+		nextInPath := filepath.Join(input, entry.Name())
+		nextOutPath := filepath.Join(output, entry.Name())
+
+		if entry.IsDir() {
+			err := processInputDir(nextInPath, nextOutPath, g)
+			if err != nil {
+				return err
+			}
+		} else {
+			inString, err := readInput(nextInPath)
+			if err != nil {
+				return err
+			}
+			if err := renderTemplate(g, inString, nextOutPath); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// == File handling ================================================
+
+func statDir(dir string) (os.FileInfo, error) {
+	si, err := os.Stat(dir)
+	if err != nil {
+		return nil, err
+	}
+	if !si.IsDir() {
+		return nil, fmt.Errorf("%s is not a directory", dir)
+	}
+	return si, nil
+}
+
+func readInputs(input string, files []string) ([]string, error) {
+	if input != "" {
+		return []string{input}, nil
+	}
+	if len(files) == 0 {
+		files = []string{"-"}
+	}
+	ins := make([]string, len(files))
+
+	for n, filename := range files {
+		inString, err := readInput(filename)
+		if err != nil {
+			return nil, err
+		}
+		ins[n] = inString
+	}
+	return ins, nil
+}
+
+func readInput(filename string) (string, error) {
+	var err error
+	var inFile *os.File
+	if filename == "-" {
+		inFile = os.Stdin
+	} else {
+		inFile, err = os.Open(filename)
+		if err != nil {
+			return "", fmt.Errorf("failed to open %s\n%v", filename, err)
+		}
+		defer checkClose(inFile, &err)
+	}
+	bytes, err := ioutil.ReadAll(inFile)
+	if err != nil {
+		err = fmt.Errorf("read failed for %s\n%v", filename, err)
+		return "", err
+	}
+	return string(bytes), nil
+}
+
+func openOutFile(filename string) (out *os.File, err error) {
+	if filename == "-" {
+		return os.Stdout, nil
+	}
+	return os.OpenFile(filename, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0644)
+}
+
+func checkClose(c io.Closer, err *error) {
```

It's not me, its the gometalinter which croaks if the error is not checked  ;-).

Well, if `*error` is nil then it's set with the error from the `Close()` so it makes a _little_ sense.